streams/src/main/java/org/apache/kafka/streams/state/internals/CachingSessionStore.java

```diff
@@ -178,27 +177,34 @@ public void put(final Windowed<Bytes> key, final byte[] value) {
 
     private void putAndMaybeForward(final ThreadCache.DirtyEntry entry, final InternalProcessorContext context) {
         final Bytes binaryKey = cacheFunction.key(entry.key());
-        final ProcessorRecordContext current = context.recordContext();
-        context.setRecordContext(entry.entry().context());
-        try {
-            final Windowed<K> key = SessionKeySchema.from(binaryKey.get(), serdes.keyDeserializer(), topic);
-            final Bytes rawKey = Bytes.wrap(serdes.rawKey(key.key()));
-            if (flushListener != null) {
-                final AGG newValue = serdes.valueFrom(entry.newValue());
-                final AGG oldValue = newValue == null || sendOldValues ?
-                    serdes.valueFrom(bytesStore.fetchSession(rawKey, key.window().start(), key.window().end())) :
-                    null;
-                if (!(newValue == null && oldValue == null)) {
+        final Windowed<Bytes> bytesKey = SessionKeySchema.from(binaryKey);
+        if (flushListener != null) {
+            final byte[] newValueBytes = entry.newValue();
+            final byte[] oldValueBytes = bytesStore.fetchSession(bytesKey.key(), bytesKey.window().start(), bytesKey.window().end());
+
+            // this is an optimization: if this key did not exist in underlying store and also not in the cache,
+            // we can skip flushing to downstream as well as writing to underlying store
+            if (newValueBytes != null || oldValueBytes != null) {
+                final Windowed<K> key = SessionKeySchema.from(bytesKey, serdes.keyDeserializer(), topic);
+                final AGG newValue = newValueBytes != null ? serdes.valueFrom(newValueBytes) : null;
+                final AGG oldValue = sendOldValues && oldValueBytes != null ? serdes.valueFrom(oldValueBytes) : null;
+                // we need to get the old values if needed, and then put to store, and then flush
+                bytesStore.put(bytesKey, entry.newValue());
```

Comment:
This is another optimization I did for window / session stores: previously we deserialize bytes to windowed<K>, and then we serialize it again to windowed<Bytes> for underlying.put, which is a bit waste.

Now what I did is `bytes -> windowed<Bytes> -> windowed<K>` where the first deser is always executed, while the second deser is executed only if new / old bytes are not null.